### ISSUE TITLE
XMLSerializer fails with System.ArgumentNullException - Blazor Webasm/.Net 5.0/Trimming

### ISSUE THREAD

**Issue by LaughingJohn** [dotnet/aspnetcore#34337]
Hi,

I'm not entirely sure if this is Blazor Webasm specific so apologies if this in the wrong place.

We have a database containing data that was previously serialized using the old XmlSerializer. In my Blazor App I need to be able to Serialize/Deserialize this data. I added code for this and it was all working locally but started failing when I published the application. This is the result of trying to deserialize:

```
fail: MyCompany.Blazor.Library.ContainerServices.ExceptionHandlerService[0]
System.InvalidOperationException: Error during deserialization
 ---> System.ArgumentNullException: ArgumentNull_Generic Arg_ParamName_Name, obj
   at System.Reflection.Emit.ModuleBuilder.GetToken(MemberInfo member, Boolean create_open_instance)
   at System.Reflection.Emit.ModuleBuilderTokenGenerator.GetToken(MemberInfo member, Boolean create_open_instance)
   at System.Reflection.Emit.ILGenerator.Emit(OpCode opcode, ConstructorInfo con)
   at System.Xml.Serialization.CodeGenerator.Call(ConstructorInfo ctor)
   at System.Xml.Serialization.XmlSerializationILGen.GenerateSerializerContract(String className, XmlMapping[] xmlMappings, Type[] types, String readerType, String[] readMethods, String writerType, String[] writerMethods, Dictionary`2 serializers)
   at System.Xml.Serialization.TempAssembly.GenerateRefEmitAssembly(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace)
   at System.Xml.Serialization.TempAssembly..ctor(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace, String location)
   at System.Xml.Serialization.XmlSerializer.GenerateTempAssembly(XmlMapping xmlMapping, Type type, String defaultNamespace, String location)
   at System.Xml.Serialization.XmlSerializer.GenerateTempAssembly(XmlMapping xmlMapping, Type type, String defaultNamespace)
   at System.Xml.Serialization.XmlSerializer..ctor(Type type, String defaultNamespace)
   at System.Xml.Serialization.XmlSerializer..ctor(Type type)
   at MyCompany.Blazor.Library.Helpers.SerializationHelper.DeserializeXml[RichPackage](String xml)
   Exception_EndOfInnerExceptionStack
   at MyCompany.Blazor.Library.Helpers.SerializationHelper.DeserializeXml[RichPackage](String xml)
   at MyCompany.Blazor.Client.Components.FmFormRichTextBox.Intialise()
   at MyCompany.Blazor.Library.ContainerServices.ExceptionHandlerService.ExecuteAndHandle(Action action)
```

Just in case it's relevant, the structure of my application is as follows:

- The main application is a Blazor web assembly app (.Net 5.0).
- The types to be serialized/de-serialized are contained in a .Net Standard 2.0 library.
- The code which actually does the serialization is in a .Net 5.0 blazor library.

The actual code looks like this:

```
        public static T DeserializeXml<T>(string xml)
        {
            try
            {
                using (var memoryStream = new MemoryStream(Encoding.UTF8.GetBytes(xml)))
                {
                    var settings = new XmlReaderSettings { IgnoreWhitespace = false };
                    var xmlReader = XmlReader.Create(memoryStream, settings);
                    var serializer = new XmlSerializer(typeof(T));

                    var obj = serializer.Deserialize(xmlReader);
                    return (T)obj;
                }
            }
            catch (Exception e)
            {
                throw new InvalidOperationException("Error during deserialization", e);
            }
        }
```
Because this was only failing after publishing I suspected the linker/trimmer and after much trial and error finally managed to get it to work by adding this to the csproj file:

```
  <ItemGroup>
    <TrimmerRootAssembly Include="System.Xml.XmlSerializer" />
  </ItemGroup>
```

As I have a workaround this is not urgent, but I wanted to make you aware in case a fix is possible and also in case it helps if someone else comes across the same issue.

Thanks,
LJ.

**Comment by LaughingJohn** [comment]
I came across this issue, which may be similar: [https://github.com/dotnet/runtime/issues/47654](https://github.com/dotnet/runtime/issues/47654)

**Comment by eerhardt** [comment]
This issue is fixed in 6.0. It doesn't reproduce for me using the latest 6.0 builds. I believe it was fixed by https://github.com/dotnet/runtime/pull/44772.

@LaughingJohn - can you try it against 6.0 and let me know if it is still an issue? Using the workaround for 5.0 seems like the right approach until 6.0 is released.

**Comment by LaughingJohn** [comment]
Thanks @eerhardt that's great. I will try it if I get chance, but may not be until we start thinking about actually moving to 6.0. Happy with the workaround for now. Feel free to close this!

**Comment by eerhardt** [comment]
Closing as this was fixed in 6.0. Feel free to re-open if this reproduces in 6.
